src/protocolsupport/protocol/packet/middleimpl/clientbound/play/v_pe/CraftingData.java

```diff
@@ -0,0 +1,143 @@
+			if (recipe instanceof ShapelessRecipe) {
+				ShapelessRecipe shapeless = (ShapelessRecipe)recipe;
+
+				expandRecipe(shapeless.getIngredients(), ingredients -> {
+					addRecipeShapeless(buffer, shapeless.getResult(), ingredients);
+				});
+			} else if(recipe instanceof ShapedRecipe) {
+				ShapedRecipe shaped = (ShapedRecipe)recipe;
+
+				expandRecipe(shaped.getIngredients(), ingredients -> {
+					addRecipeShaped(buffer, shaped.getResult(), shaped.getWidth(), shaped.getHeight(), ingredients);
+				});
+			} else if(recipe instanceof SmeltingRecipe) {
+				SmeltingRecipe smelting = (SmeltingRecipe)recipe;
+				expandRecipe(new Ingredient[]{smelting.getIngredient()}, ingredients -> {
+					addRecipeFurnace(buffer, smelting.getResult(), ingredients.get(0));
+				});
+			}
+		}
+		ClientBoundPacketData craftPacket = ClientBoundPacketData.create(PEPacketIDs.CRAFTING_DATA);
+		VarNumberSerializer.writeVarInt(craftPacket, recipesWritten);
+		craftPacket.writeBytes(buffer);
+		buffer.release();
+		return RecyclableSingletonList.create(craftPacket);
+	}
+
+	protected void addRecipeShaped(ByteBuf to, NetworkItemStack output, int width, int height, List<NetworkItemStack> required) {
+		VarNumberSerializer.writeSVarInt(to, 1); //recipe type
+		VarNumberSerializer.writeSVarInt(to, width);
+		VarNumberSerializer.writeSVarInt(to, height);
+		for (NetworkItemStack stack : required) {
+			ItemStackSerializer.writeItemStack(to, connection.getVersion(), I18NData.DEFAULT_LOCALE, stack, stack == null || stack.getLegacyData() != -1);
+		}
+		VarNumberSerializer.writeVarInt(to, 1); // result item count (PC only supports one itemstack output, so hardcoded to 1)
+		ItemStackSerializer.writeItemStack(to, connection.getVersion(), I18NData.DEFAULT_LOCALE, output, true);
+		MiscSerializer.writeUUID(to, connection.getVersion(), UUID.nameUUIDFromBytes(to.array()));
+		recipesWritten++;
+	}
+
+	protected void addRecipeShapeless(ByteBuf to, NetworkItemStack output, List<NetworkItemStack> required) {
+		VarNumberSerializer.writeSVarInt(to, 0); //recipe type
+		VarNumberSerializer.writeVarInt(to, required.size());
+		for (NetworkItemStack stack : required) {
+			ItemStackSerializer.writeItemStack(to, connection.getVersion(), I18NData.DEFAULT_LOCALE, stack, stack == null || stack.getLegacyData() != -1);
+		}
+		VarNumberSerializer.writeVarInt(to, 1); // result item count (PC only supports one itemstack output, so hardcoded to 1)
+		ItemStackSerializer.writeItemStack(to, connection.getVersion(), I18NData.DEFAULT_LOCALE, output, true);
+		MiscSerializer.writeUUID(to, connection.getVersion(), UUID.nameUUIDFromBytes(to.array()));
+		recipesWritten++;
+	}
+
+	protected void addRecipeFurnace(ByteBuf to, NetworkItemStack output, NetworkItemStack input) {
+		int peCombinedId = PEItems.getPECombinedIdByModernId(input.getTypeId());
+		if (PEItems.getDataFromPECombinedId(peCombinedId) == 0) {
+			VarNumberSerializer.writeSVarInt(to, 2); //recipe type
```

Comment:
Please extract the recipe type to public static final int constant. This will make it clearer.